code/ui/InventorySelection.cs

```diff
@@ -105,42 +109,63 @@ public override void Tick()
             {
                 return;
             }
 
-            int selectedInventoryIndex = SlotPressInput(input);
+            ICarriableItem activeCarriable = Local.Pawn.ActiveChild as ICarriableItem;
 
-            // no button pressed, but maybe scrolled
-            if (selectedInventoryIndex == 0)
+            int keyboardIndexPressed = GetKeyboardNumberPressed(input);
+            if (keyboardIndexPressed != 0)
             {
-                if (input.MouseWheel != 0)
-                {
-                    int nextSlot = inventory.List.IndexOf(Local.Pawn.ActiveChild) - input.MouseWheel;
-                    nextSlot = inventory.NormalizeSlotIndex(nextSlot, inventory.Count() - 1);
+                // Find all weapons of hold type that equals the number the user pressed on their keyboard.
+                List<InventorySlot> weaponsOfHoldType = _children.ConvertAll(p => p as InventorySlot).Where(
+                    (slot) => (int) slot.Carriable.HoldType == keyboardIndexPressed).ToList();
 
-                    input.ActiveChild = inventory.GetSlot(nextSlot);
+                // Check to see if a user is currently holding a weapon which is of the same hold type.
+                bool isActiveCarriableOfWeaponType = weaponsOfHoldType.Any((slot) =>
+                    slot.Carriable.Name == activeCarriable?.Name);
 
-                    input.MouseWheel = 0;
+                if (activeCarriable == null || !isActiveCarriableOfWeaponType)
+                {
+                    // The user isn't holding a weapon, or is holding a weapon of a different hold type.
+                    // We can just select the first weapon.
+                    input.ActiveChild = weaponsOfHoldType.FirstOrDefault()?.Carriable as Entity;
+                }
+                else
+                {
+                    // The user is holding a weapon of the currently selected hold type.
+                    // Increment the index using GetNextWeaponIndex.
+                    int activeCarriableIndex = weaponsOfHoldType.FindIndex((slot) => slot.Carriable.Name == activeCarriable?.Name);
+                    input.ActiveChild = weaponsOfHoldType[GetNextWeaponIndex(activeCarriableIndex, weaponsOfHoldType.Count)].Carriable as Entity;
                 }
-
-                return;
             }
 
-            // corresponding key pressed
-            int nextInventorySlot = inventory.GetNextSlot((HoldType) selectedInventoryIndex);
-
-            if (nextInventorySlot != -1)
+            int mouseWheelIndex = Input.MouseWheel;
+            if (mouseWheelIndex != 0)
             {
-                input.ActiveChild = inventory.GetSlot(nextInventorySlot);
+                int activeCarriableIndex = _children.FindIndex((p) =>
+                    p is InventorySlot slot && slot.Carriable.Name == activeCarriable?.Name);
+
+                input.ActiveChild = (_children[NormalizeSlotIndex(mouseWheelIndex + activeCarriableIndex, _children.Count - 1)]
```

Comment:
```suggestion
                input.ActiveChild = (_children[NormalizeSlotIndex(-mouseWheelIndex + activeCarriableIndex, _children.Count - 1)]
```

I'd switch like you're scrolling: down to select downwards, up to select upwards